middleware/src/middleware_test.go

```diff
@@ -1,15 +1,14 @@
-package middleware_test
+package middleware
```

Comment:
This is okay, but a few suggestions if you like them (we iterated those guidelines in my former company):

- If you test the public interface of the package without accessing internals, keep `_test`, which does not allow accessing the unexported stuff
- If you do explicitly want to test unexported functions. consider splitting those out into their own package. If unreasonable to do that, name the file `middleware_internal_test.go` and test only internals there, and keep the the public tests in `middleware_test.go`.
- If you only need  to access like one or two little things which are unexported, or you need to set some test mock function which can only be done via internals, you can make a file `middleware_export_test.go` and export your unexported things. See this little tutorial: https://medium.com/@robiplus/golang-trick-export-for-test-aa16cbd7b8cd

The benefit of this is that it's very clear what part of the code is tested. Imho testing only the public interface is best (of course exceptions exist), and testing internals is an indication that the internals are too complicated and need to be split.

Reply:
Thank you for sharing this!